rule_packages/c/IO4.json

```diff
@@ -0,0 +1,71 @@
+{
+  "CERT-C": {
+    "FIO45-C": {
+      "properties": {
+        "obligation": "rule"
+      },
+      "queries": [
+        {
+          "description": "TOCTOU race conditions when accessing files can lead to vulnerability.",
+          "kind": "problem",
+          "name": "Avoid TOCTOU race conditions while accessing files",
+          "precision": "high",
+          "severity": "error",
+          "short_name": "ToctouRaceConditionsWhileAccessingFiles",
+          "tags": [
+            "correctness",
+            "security"
+          ],
+          "implementation_scope": {
+            "description": "The query is limited to the specific class of TOCTOU race conditions."
+          }
+        }
+      ],
+      "title": "Avoid TOCTOU race conditions while accessing files"
+    },
+    "FIO47-C": {
+      "properties": {
+        "obligation": "rule"
+      },
+      "queries": [
+        {
+          "description": "Use valid conversion specifier.",
+          "kind": "problem",
+          "name": "Use valid format strings",
+          "precision": "high",
+          "severity": "error",
+          "short_name": "UseValidSpecifiers",
+          "tags": [
+            "correctness",
+            "security"
+          ]
+        },
+        {
+          "description": "Use correct number of format arguments.",
+          "kind": "problem",
+          "name": "Use correct number fo arguments",
+          "precision": "high",
+          "severity": "error",
+          "short_name": "WrongNumberOfFormatArguments",
+          "tags": [
+            "correctness",
+            "security"
+          ]
+        },
+        {
+          "description": "Wrong type format arguments.",
+          "kind": "problem",
+          "name": "Wrong type format arguments",
+          "precision": "high",
+          "severity": "error",
+          "short_name": "WrongTypeFormatArguments",
+          "tags": [
+            "correctness",
+            "security"
+          ]
+        }
+      ],
+      "title": "Use valid format strings"
```

Comment:
not sure if we want to add an implementation notes section that comments on this part of the rule: "Format strings are usually string literals specified at the call site, but they need not be. However, they should not contain tainted values" since it seems clearly fair that this rule only checks string literals specified at the call site, and doesnt consider flow/string composition since thats handled in FIO30-c

Reply:
being pretty picky on the scope of the implementation I know but do we also want to comment on :

"The formatted input functions (fscanf() and related functions) use similarly specified format strings and impose similar restrictions on their format strings and arguments."

I think `FormattingFunction` is only the `printf` family, not `scanf` but I think thats fine as long as we mention it?

Reply:
ok, only one more possible suggestion on this (the rule has a lot of little considerations haha), do we wanna also mention that we dont look at the no effect cases? (which again I think makes sense since they arent invalid)

Reply:
"Do you mean that the N/E combinations could be reported as warning as the length modifier should be omitted?" yeah that was what I had in mind, 
but actually on second consideration lets not, as those are valid and the rule is more about finding invalid rather than valid but have no effect.